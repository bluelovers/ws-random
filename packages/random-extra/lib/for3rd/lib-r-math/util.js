Object.defineProperty(exports, "__esModule", {
    value: !0
});

const libRmath = require("lib-r-math.js");

function fakeLibRMathRng(e) {
    return {
        unif_rand(t) {
            if (t > 1) {
                let r = [];
                for (;t--; ) r[t] = e();
                return r;
            }
            return e();
        }
    };
}

exports.libRmath = libRmath, exports.fakeLibRMathRng = fakeLibRMathRng;
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIjAiXSwibmFtZXMiOlsiT2JqZWN0IiwiZGVmaW5lUHJvcGVydHkiLCJleHBvcnRzIiwidmFsdWUiLCJsaWJSbWF0aCIsInJlcXVpcmUiLCJmYWtlTGliUk1hdGhSbmciLCJmbiIsIltvYmplY3QgT2JqZWN0XSIsIm4iLCJhIl0sIm1hcHBpbmdzIjoiQUFHQUEsT0FBT0MsZUFBZUMsU0FBUztJQUFnQkMsUUFBTzs7O0FBQ3RELE1BQU1DLFdBQVdDLFFBQVE7O0FBRXpCLFNBQVNDLGdCQUFnQkM7SUFDckI7UUFDSUMsVUFBVUM7WUFDTixJQUFJQSxJQUFJLEdBQUc7Z0JBQ1AsSUFBSUM7Z0JBQ0osTUFBT0QsT0FDSEMsRUFBRUQsS0FBS0Y7Z0JBRVgsT0FBT0c7O1lBRVgsT0FBT0g7Ozs7O0FBWG5CTCxRQUFRRSxXQUFXQSxVQWVuQkYsUUFBUUksa0JBQWtCQSJ9