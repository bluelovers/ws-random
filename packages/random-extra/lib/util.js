Object.defineProperty(exports, "__esModule", {
    value: !0
});

const hashSum = require("hash-sum"), _nanoid = require("nanoid/non-secure"), _pkg = require("../package.json"), const_1 = require("./util/const"), _GLOBAL = (0, 
eval)("this");

function randomSeedNum() {
    return _MathRandom() * const_1.MATH_POW_2_32 + _MathRandom();
}

function randomSeedStr() {
    return [ _nanoid(), hashSum(_pkg.name), hashSum(_pkg.version), Date.now(), floatToString(_MathRandom()) ].join("_");
}

function getClass(n, t, ...e) {
    let r;
    return r = t instanceof n ? t.__proto__.constructor : n;
}

function cloneClass(n, t, ...e) {
    let r;
    return new (getClass(n, t, ...e))(...e);
}

function floatToString(n) {
    let t = Math.floor(n), e, r = String(n - t).split(".")[1];
    return String(t) + (r ? "." + r : "");
}

function expectInDelta(n, t, e = .05) {
    return t - e <= n && n <= t + e;
}

function hashArgv(n) {
    return String(n.join(";"));
}

function isNum(n) {
    return n === +n;
}

function isInt(n) {
    return n === (0 | n);
}

function isFloat(n) {
    return n === +n && n !== (0 | n);
}

function isUnset(n) {
    return void 0 === n || null === n;
}

function array_unique_unsafe(n) {
    return n.filter((n, t, e) => e.indexOf(n) === t);
}

exports._MathRandom = _MathRandom = Math.random, exports.randomSeedNum = randomSeedNum, 
exports.randomSeedStr = randomSeedStr, exports.getClass = getClass, exports.cloneClass = cloneClass, 
exports.floatToString = floatToString, exports.expectInDelta = expectInDelta, exports.hashArgv = hashArgv, 
exports.isNum = isNum, exports.isInt = isInt, exports.isFloat = isFloat, exports.isUnset = isUnset, 
exports.array_unique_unsafe = array_unique_unsafe, Object.freeze(exports);
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIjAiXSwibmFtZXMiOlsiT2JqZWN0IiwiZGVmaW5lUHJvcGVydHkiLCJleHBvcnRzIiwidmFsdWUiLCJoYXNoU3VtIiwicmVxdWlyZSIsIl9uYW5vaWQiLCJfcGtnIiwiY29uc3RfMSIsIl9HTE9CQUwiLCJldmFsIiwicmFuZG9tU2VlZE51bSIsIl9NYXRoUmFuZG9tIiwiTUFUSF9QT1dfMl8zMiIsInJhbmRvbVNlZWRTdHIiLCJuYW1lIiwidmVyc2lvbiIsIkRhdGUiLCJub3ciLCJmbG9hdFRvU3RyaW5nIiwiam9pbiIsImdldENsYXNzIiwiUk5HQ2xhc3MiLCJ0aGlzQXJndiIsImFyZ3YiLCJvIiwiX19wcm90b19fIiwiY2xvbmVDbGFzcyIsIm4iLCJpbnQiLCJNYXRoIiwiZmxvb3IiLCJmbG9hdCIsInMiLCJTdHJpbmciLCJzcGxpdCIsImV4cGVjdEluRGVsdGEiLCJhY3R1YWwiLCJleHBlY3RlZCIsImRlbHRhIiwiaGFzaEFyZ3YiLCJhcmdzIiwiaXNOdW0iLCJpc0ludCIsImlzRmxvYXQiLCJpc1Vuc2V0IiwiYXJyYXlfdW5pcXVlX3Vuc2FmZSIsImFyciIsImZpbHRlciIsInYiLCJpIiwiaW5kZXhPZiIsInJhbmRvbSIsImZyZWV6ZSJdLCJtYXBwaW5ncyI6IkFBR0FBLE9BQU9DLGVBQWVDLFNBQVM7SUFBZ0JDLFFBQU87OztBQUN0RCxNQUFNQyxVQUFVQyxRQUFRLGFBQ2xCQyxVQUFVRCxRQUFRLHNCQUNsQkUsT0FBT0YsUUFBUSxvQkFDZkcsVUFBVUgsUUFBUSxpQkFPbEJJLFdBQVU7QUFBSUMsTUFBTTs7QUFHMUIsU0FBU0M7SUFDTCxPQUFRQyxnQkFBZ0JKLFFBQVFLLGdCQUFpQkQ7OztBQU1yRCxTQUFTRTtJQUNMLFNBQ0lSLFdBQ0FGLFFBQVFHLEtBQUtRLE9BQ2JYLFFBQVFHLEtBQUtTLFVBQ2JDLEtBQUtDLE9BQ0xDLGNBQWNQLGlCQUNoQlEsS0FBSzs7O0FBTVgsU0FBU0MsU0FBU0MsR0FBVUMsTUFBYUM7SUFDckMsSUFBSUM7SUFRSixPQUxJQSxJQUZBRixhQUFvQkQsSUFFZkMsRUFBU0csVUFBcUIsY0FHL0JKOzs7QUFRWixTQUFTSyxXQUFXTCxHQUFVQyxNQUFhQztJQUN2QyxJQUFJQztJQUNKLE9BQU8sS0FEQ0osU0FBU0MsR0FBVUMsTUFBYUMsR0FDakMsSUFBU0E7OztBQUdwQixTQUFTTCxjQUFjUztJQUNuQixJQUFJQyxJQUFNQyxLQUFLQyxNQUFNSCxJQUNqQkksR0FDQUMsSUFBSUMsT0FESU4sSUFBSUMsR0FFWE0sTUFBTSxLQUFLO0lBQ2hCLE9BQU9ELE9BQU9MLE1BQVFJLElBQUksTUFBTUEsSUFBSTs7O0FBVXhDLFNBQVNHLGNBQWNDLEdBQVFDLEdBQVVDLElBQVE7SUFDN0MsT0FBT0QsSUFBV0MsS0FBU0YsS0FBVUEsS0FBVUMsSUFBV0M7OztBQUc5RCxTQUFTQyxTQUFTQztJQUNkLE9BQU9QLE9BQU9PLEVBQUtyQixLQUFLOzs7QUFHNUIsU0FBU3NCLE1BQU1kO0lBQ1gsT0FBT0EsT0FBT0E7OztBQU1sQixTQUFTZSxNQUFNZjtJQUNYLE9BQU9BLE9BQVcsSUFBSkE7OztBQUdsQixTQUFTZ0IsUUFBUWhCO0lBQ2IsT0FBT0EsT0FBT0EsS0FBS0EsT0FBVyxJQUFKQTs7O0FBRzlCLFNBQVNpQixRQUFRakI7SUFDYixZQUFvQixNQUFOQSxLQUEyQixTQUFOQTs7O0FBTXZDLFNBQVNrQixvQkFBb0JDO0lBQ3pCLE9BQU9BLEVBQUlDLE9BQU8sQ0FBQ0MsR0FBR0MsR0FBR0gsTUFBUUEsRUFBSUksUUFBUUYsT0FBT0M7OztBQXZGeERoRCxRQUFRVSxjQUFjQSxjQUFja0IsS0FBS3NCLFFBSXpDbEQsUUFBUVMsZ0JBQWdCQTtBQWF4QlQsUUFBUVksZ0JBQWdCQSxlQWV4QlosUUFBUW1CLFdBQVdBLFVBUW5CbkIsUUFBUXlCLGFBQWFBO0FBUXJCekIsUUFBUWlCLGdCQUFnQkEsZUFXeEJqQixRQUFRa0MsZ0JBQWdCQSxlQUl4QmxDLFFBQVFzQyxXQUFXQTtBQUluQnRDLFFBQVF3QyxRQUFRQSxPQU9oQnhDLFFBQVF5QyxRQUFRQSxPQUloQnpDLFFBQVEwQyxVQUFVQSxTQUlsQjFDLFFBQVEyQyxVQUFVQTtBQU9sQjNDLFFBQVE0QyxzQkFBc0JBLHFCQUU5QjlDLE9BQU9xRCxPQUFPbkQifQ==