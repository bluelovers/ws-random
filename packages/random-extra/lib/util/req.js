function tryRequire(e) {
    return require(e);
}

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.tryRequire = tryRequire, Object.freeze(exports);
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIjAiXSwibmFtZXMiOlsidHJ5UmVxdWlyZSIsIm5hbWUiLCJyZXF1aXJlIiwiT2JqZWN0IiwiZGVmaW5lUHJvcGVydHkiLCJleHBvcnRzIiwidmFsdWUiLCJmcmVlemUiXSwibWFwcGluZ3MiOiJBQUNBLFNBQVNBLFdBQVdDO0lBQ2hCLE9BQU9DLFFBQVFEOzs7QUFGbkJFLE9BQU9DLGVBQWVDLFNBQVM7SUFBZ0JDLFFBQU87SUFJdERELFFBQVFMLGFBQWFBLFlBRXJCRyxPQUFPSSxPQUFPRiJ9