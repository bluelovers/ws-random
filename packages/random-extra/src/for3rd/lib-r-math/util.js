/**
 * Created by user on 2018/11/19/019.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const libRmath = require("lib-r-math.js");
exports.libRmath = libRmath;
function fakeLibRMathRng(fn) {
    return {
        unif_rand(n) {
            if (n > 1) {
                let a = [];
                while (n--) {
                    a[n] = fn();
                }
                return a;
            }
            return fn();
        },
    };
}
exports.fakeLibRMathRng = fakeLibRMathRng;
