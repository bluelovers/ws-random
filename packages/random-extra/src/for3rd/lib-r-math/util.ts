/**
 * Created by user on 2018/11/19/019.
 */

import * as libRMath from 'lib-r-math.js';
export { libRMath }

export function fakeLibRMathRng(fn: () => number)
{
	return {
		unif_rand(n?: number)
		{
			if (n > 1)
			{
				let a = [];
				while (n--)
				{
					a[n] = fn()
				}
				return a;
			}

			return fn()
		},
	} as libRMath.IRNG
}
