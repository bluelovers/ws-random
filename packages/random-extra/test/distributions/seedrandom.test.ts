/**
 * Created by user on 2018/11/16/016.
 */

/// <reference types="mocha" />
/// <reference types="benchmark" />
/// <reference types="chai" />
/// <reference types="node" />

import RNG from '../../src/rng';
import random, { Random } from '../../src/'
import seedrandomOrigin from 'seedrandom';
import RNGSeedRandom from '../../src/generators/seedrandom';
import { tryRequire } from '../../src/util/req';
import checkTypesMatchers from '../jest/type';

expect.extend(checkTypesMatchers);

// @ts-ignore
describe(`'hello.' 0.9282578795792454`, () =>
{
	let seed = 'hello.';
	const expected = 0.9282578795792454;

	const count = 10000;
	let val: number;

	it(`RNGSeedRandom`, () =>
	{
		expect(RNGSeedRandom.create('hello.', null)).toBeInstanceOf(RNG);
	});

	// @ts-ignore
	it(`all seedrandom(seed) should has same value`, () =>
	{
		val = [
			random.newUse(seedrandomOrigin('hello.')),
			random.newUse('seedrandom', 'hello.', null),
			random.newUse(RNGSeedRandom.create('hello.', null)),
		].reduce(function (pre, rnd, idx)
		{
			let value = rnd.float()

			if (pre !== null)
			{
				// idx.toString()
				expect(value).toEqual(pre);
				expect(value).toBeCloseToWithDelta(pre, 0.000001);
				expect(value).toBeCloseToWithDelta(pre + 1 - 1, pre + 1 - 1);
			}

			return value
		}, expected)
	});

	it(`rnd.float() = ${expected}`, () =>
	{
		expect(val).toEqual(expected);
	});

});

describe(`test sub seedrandom`, () =>
{
	const subs = [
		'alea', 'tychei', 'xor128', 'xor4096', 'xorshift7', 'xorwow',
	];

	const methods = ['quick', 'int32', 'double'];

	subs.forEach(function (libName)
	{
		it(`RNGSeedRandom.createLib('${libName}', seed, opts)`, () =>
		{
			let seed = 'hello.';
			let opts = null;
			const r = RNGSeedRandom.createLib(libName, seed, opts);

			// @ts-ignore
			const _seedrandom = r._seedrandom;

			let s = _seedrandom(seed, opts);

			//console.log(libName, s.quick(), r.next());

			expect(s).toContainAllKeys(methods);
			expect(r.next()).not.toBeNaN();

			let v = r.next();

			expect(v).toBeFloat();
			expect(v).toBeGreaterThan(0)
			expect(v).toBeLessThan(1)

			expect(_seedrandom).toEqual(tryRequire('seedrandom')[libName])

			expect(_seedrandom).toBeInstanceOf(Function);
		})
	});

	subs.forEach(function (libName)
	{
		it(`random.newUse('seedrandom', seed, opts, '${libName}')`, () =>
		{
			let seed = 'hello.';
			let opts = null;
			const r = random.newUse('seedrandom', seed, opts, libName);

			// @ts-ignore
			const _seedrandom = r.rng._seedrandom;

			let s = _seedrandom(seed, opts);

			//console.log(libName, s.quick(), r.next());

			expect(s).toContainAllKeys(methods);
			expect(r.next()).not.toBeNaN();

			let v = r.next();

			expect(v).toBeFloat();
			expect(v).toBeGreaterThan(0)
			expect(v).toBeLessThan(1)

			expect(_seedrandom).toEqual(tryRequire('seedrandom')[libName])

			expect(_seedrandom).toBeInstanceOf(Function);
		})
	});

})

