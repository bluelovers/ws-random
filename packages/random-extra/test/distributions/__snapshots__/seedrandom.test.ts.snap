// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test sub seedrandom RNGSeedRandom.createLib('alea', seed, opts) 1`] = `0.8297006865032017`;

exports[`test sub seedrandom RNGSeedRandom.createLib('tychei', seed, opts) 1`] = `0.794256798690185`;

exports[`test sub seedrandom RNGSeedRandom.createLib('xor128', seed, opts) 1`] = `0.6919899098575115`;

exports[`test sub seedrandom RNGSeedRandom.createLib('xor4096', seed, opts) 1`] = `0.035834695445373654`;

exports[`test sub seedrandom RNGSeedRandom.createLib('xorshift7', seed, opts) 1`] = `0.017942105419933796`;

exports[`test sub seedrandom RNGSeedRandom.createLib('xorwow', seed, opts) 1`] = `0.5642471052706242`;

exports[`test sub seedrandom random.newUse('seedrandom', seed, opts, 'alea') 1`] = `0.8297006865032017`;

exports[`test sub seedrandom random.newUse('seedrandom', seed, opts, 'tychei') 1`] = `0.794256798690185`;

exports[`test sub seedrandom random.newUse('seedrandom', seed, opts, 'xor128') 1`] = `0.6919899098575115`;

exports[`test sub seedrandom random.newUse('seedrandom', seed, opts, 'xor4096') 1`] = `0.035834695445373654`;

exports[`test sub seedrandom random.newUse('seedrandom', seed, opts, 'xorshift7') 1`] = `0.017942105419933796`;

exports[`test sub seedrandom random.newUse('seedrandom', seed, opts, 'xorwow') 1`] = `0.5642471052706242`;
